### PR TITLE
feat(http): dispose EventEmitter on XHRConnection abort

### PR DESCRIPTION
**Pull request by PatrickJS**
dispose EventEmitter on xhr abort

[Review on Reviewable](https://reviewable.io/reviews/angular/angular/2668)

### CODE REVIEW

File: modules/angular2/src/http/backends/xhr_backend.ts [hunk 1]
```diff
@@ -45,7 +45,10 @@ export class XHRConnection implements Connection {
   /**
    * Calls abort on the underlying XMLHttpRequest.
    */
-  dispose(): void { this._xhr.abort(); }
+  dispose(): void {
+    this._xhr.abort();
+    this.response.dispose();
```
Comment (vicb) [inline modules/angular2/src/http/backends/xhr_backend.ts]: could you please add a test ?

**Comment by gdi2290** [comment]
This PR is blocked.
As far as I know there is no way for me to correctly delegate `EventEmitter`'s `dispose` to `XHRConnection`'s `dispose` without creating an `Observable` during `XHRBackend#createConnection`.

If I was using RxJS I would create an Observable on each `createConnection` that will manage the dispose
```typescript
export class XHRBackend implements ConnectionBackend {
  constructor(private _browserXHR: BrowserXhr, private _baseResponseOptions: ResponseOptions) {}
  createConnection(request: Request): XHRConnection {
    return Rx.Observable.create(obs => {
        var connection = new XHRConnection(request, this._browserXHR, this._baseResponseOptions);
        var disposed = false;
        ObservableWrapper.subscribe(
          connection,
          res => (!disposed) && obs.onNext(res),
          err => (!disposed) && obs.onError(err),
          _   => (!disposed) && obs.onComplete()
        );

        return function dispose() {
          disposed = true;
          connection.dispose(); // xhr.abort
          connection.response.dispose(); // dispose EventEmitter
        };
    });
  }
}
```
I'm not sure how to do this with the `EventEmitter` and `Observable` from `facade/async`

File: modules/angular2/src/http/backends/xhr_backend.ts [hunk 2]
```diff
@@ -51,7 +51,10 @@ export class XHRConnection implements Connection {
   /**
    * Calls abort on the underlying XMLHttpRequest.
    */
-  dispose(): void { this._xhr.abort(); }
+  dispose(): void {
+    this._xhr.abort();
+    ObservableWrapper.dispose(this.response);
```
Comment (gdi2290) [inline modules/angular2/src/http/backends/xhr_backend.ts]: I don't have access to the subscription here so this won't work

File: modules/angular2/test/http/backends/xhr_backend_spec.ts [hunk 3]
```diff
@@ -22,11 +22,26 @@ import {BaseResponseOptions, ResponseOptions} from 'angular2/src/http/base_respo
 import {ResponseTypes} from 'angular2/src/http/enums';
 
 var abortSpy;
+var observerSpy;
 var sendSpy;
 var openSpy;
 var addEventListenerSpy;
 var existingXHRs = [];
 
+class SpyObserver extends SpyObject {
```
Comment (gdi2290) [inline modules/angular2/test/http/backends/xhr_backend_spec.ts]: kool story bro

File: modules/angular2/test/http/backends/xhr_backend_spec.ts [hunk 4]
```diff
@@ -22,11 +22,26 @@ import {BaseResponseOptions, ResponseOptions} from 'angular2/src/http/base_respo
...
+  next: Function;
+  throw: Function;
+  return: Function;
+  dispose: Function;
+  constructor() {
+    super();
+    this.next = this.spy('next');
+    this.throw = this.spy('throw');
+    this.return = this.spy('return');
+    this.dispose = this.spy('dispose');
```
Comment (gdi2290) [inline modules/angular2/test/http/backends/xhr_backend_spec.ts]: `dispose` is returned on a subscription not the Observable instance

**Comment by gdi2290** [comment]
oops I was cleaning up old branches and accidently deleted this branch

**Comment by jeffbcross** [comment]
Let's block this on #2794

**Comment by gdi2290** [comment]
having an `ObservableWrapper.create` and being able to use RxNext would be awesome

**Comment by tbosch** [comment]
`EventEmitter` will go away very soon. Jeff is working on integrating Rx.next into Angular and once this is done this PR won't be necessary.

**Pull request closed without merging**